fix(lvm): change calling operation to avoid garbage in json object

kro-cat opened the pull request:
**Why is this PR required? What issue does it fix?**:
The `vgs` command may print non-critical warnings to STDERR. Warnings may not necessarily result in a failure return code, which allows the program to continue with marshalling the JSON-formatted output. Combining this stream with STDIN will cause the next step at `decodeVgsJSON()` to fail due to garbage mixed in the JSON.

**What this PR does?**:
Drop the STDERR stream contents from `output` to avoid JSON mangling.

**Does this PR require any upgrade changes?**:

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Any additional information for your reviewer?** :
Volume groups created on the host may not necessarily be compatible with the lvm-driver container. This PR simply introduces a fix to isolate any warnings generated by `vgs`, but this won't fix any issues with incompatible volume groups: there may be additional steps required of the system's administrator in order to ensure his or her configuration will work with lvm-localpv. Printing `vgs`'s STDERR to the log should provide enough diagnostic information for the administrator to perform any additional steps, if needed.


**Checklist:**
- [x] Fixes #247, #249
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [x] Has the change log section been updated?
- [ ] Commit has unit tests
- [x] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

kro-cat committed:
fix(lvm): change calling operation to avoid garbage in json object

Drop the STDERR stream contents from `output` to avoid JSON mangling.
The `vgs` command may print non-critical warnings to STDERR. Warnings
may not necessarily result in a failure return code, which allows the
program to continue with marshalling the JSON-formatted output.
Combining this stream with STDIN will cause the next step at
`decodeVgsJSON()` to fail due to garbage mixed in the JSON.

Fixes #247

Signed-off-by: Kara <33235324+kro-cat@users.noreply.github.com>

kro-cat committed:
fix(lvm): add PR changelog - suppress STDERR when calling vgs

Add PR changelog

Signed-off-by: Kara <33235324+kro-cat@users.noreply.github.com>

kro-cat committed:
refactor(lvm): refactor exec to separate function, add log verbosity

Move exec code into separate function which returns STDOUT and STDERR
streams separately. This is to facilitate future occurences of that same
pattern within lvm_util.go . Additionally, add log messages to assist with
debugging lvm and json.

Signed-off-by: Kara <33235324+kro-cat@users.noreply.github.com>

kro-cat committed:
feat(ci): integrate test to simulate foreign lvm systemid

This test aims to determine whether lvm-driver is capable of operating
with a volume group whose systemid is foreign to the kubernetes host
system.
The test ensures the program in pkg/lvm/lvm_utils.go is capable of
operating in a foreign lvm environment. This is important when the
kubernetes host system has a different lvm configuration than the
lvm-driver container. Differences in configuration may cause unforseen
consequences when the host machine provisions a volume group which is to
be consumed by the lvm-driver container.
Additionally, this feature includes teardown code in the form of the
cleanup functions which aims to aid with cleaning up ci resources in a
local development environment. To make the cleanup actions compatible
with the original operation of the ci-test.sh script, the feature is
only enabled through the use of environment variables.
While this commit fixes issue #249, it's intent was as an integration
test for issue #247.

Fixes #249

Signed-off-by: Kara <33235324+kro-cat@users.noreply.github.com>

kro-cat commented:
First PR. Please let me know if there's anything I should change or if I need to do anything else to help.

kro-cat committed:
feat(ci): clean up volume groups when applicable

To avoid volume group caching, explicitly remove volume groups (and
their pvs) before removing the disk and loopback device.

Signed-off-by: Kara <33235324+kro-cat@users.noreply.github.com>

Abhinandan-Purkait commented:
Hi @kro-cat Can you please go through and resolve the [shellcheck] lint errors in files section. Thanks

kro-cat committed:
fix(ci): use -n instead of ! -z, change last statement to return true

Fix linter errors: use -n instead of ! -z in test statements. Last
statement in file is used as the return code of the file, so it must
return true; change `[ ! -z ... ] && ...` into `[ -z ... ] || ...` to
achieve this.

Signed-off-by: kro <33235324+kro-cat@users.noreply.github.com>

dnugmanov commented:
It will be great to fix also `pvs`, `lvs` - `ListLVMPhysicalVolume()`, `ListLVMLogicalVolume()`

kro-cat commented:
@dnugmanov Good catch, I'll refactor those as well

Abhinandan-Purkait commented:
> @dnugmanov Good catch, I'll refactor those as well

@kro-cat Let's take that refactor on a different PR.

kro-cat commented:
@Abhinandan-Purkait got it. Should I open a new issue for it, or is that not required for a refactor?

abhilashshetty04 commented:
@kro-cat , New issue might not be required if you explain all changes in PR

Abhinandan-Purkait reviewed (approved):
LGTM. Thanks

Abhinandan-Purkait commented:
@kro-cat We hope all your commits are in, so are we good to merge now?

kro-cat commented:
@Abhinandan-Purkait All commits are in, we're good to merge. :)
